apps/sms/views/shared/js/utils.js

```diff
@@ -33,40 +33,93 @@
     },
-    getHeaderDate: function ut_giveHeaderDate(time) {
-      var _ = navigator.mozL10n.get;
+    setHeaderDate: function ut_setHeaderDate({time, element, withTime}) {
+      var options, formatter;
       var today = Utils.getDayDate(Date.now());
       var otherDay = Utils.getDayDate(time);
       var dayDiff = (today - otherDay) / 86400000;
       this.date.shared.setTime(+time);
 
       if (isNaN(dayDiff)) {
-        return _('incorrectDate');
+        element.setAttribute('data-l10n-id', 'incorrectDate');
+        return;
       }
 
       if (dayDiff < 0) {
         // future time
-        return this.date.format.localeFormat(
-          this.date.shared, '%x'
-        );
+        options = {
+          month: '2-digit',
+          day: '2-digit',
+          year: 'numeric',
+        }; 
+        if (withTime) {
+          options.hour12 = navigator.mozHour12;
+          options.hour = 'numeric';
+          options.minute = 'numeric';
+        }
+        formatter = new Intl.DateTimeFormat(navigator.languages, options);
+        element.removeAttribute('data-l10n-id');
+        element.textContent = formatter.format(this.date.shared);
+        return;
       }
 
-      return dayDiff === 0 && _('today') ||
-        dayDiff === 1 && _('yesterday') ||
-        dayDiff < 6 && this.date.format.localeFormat(this.date.shared, '%A') ||
-        this.date.format.localeFormat(this.date.shared, '%x');
+      if (dayDiff === 0) {
+        if (withTime) {
+          navigator.mozL10n.setAttributes(
+            element, 'todayWithTime', {
+              time: this.getFormattedHour(this.date.shared)
+            });
+        } else {
+          element.setAttribute('data-l10n-id', 'today');
+        }
+      } else if (dayDiff === 1) {
+        if (withTime) {
+          navigator.mozL10n.setAttributes(
+            element, 'yesterdayWithTime', {
+              time: this.getFormattedHour(this.date.shared)
+            });
+        } else {
+          element.setAttribute('data-l10n-id', 'yesterday');
+        }
+      } else if (dayDiff < 6) {
+        options = {
+          weekday: 'long'
+        };
+        if (withTime) {
+          options.hour12 = navigator.mozHour12;
+          options.hour = 'numeric';
+          options.minute = 'numeric';
+        }
+        element.removeAttribute('data-l10n-id');
+        formatter = new Intl.DateTimeFormat(navigator.languages, options);
+        element.textContent = formatter.format(this.date.shared);
+      } else {
+        options = {
+          month: '2-digit',
+          day: '2-digit',
+          year: 'numeric',
+        };
+        if (withTime) {
+          options.hour12 = navigator.mozHour12;
+          options.hour = 'numeric';
+          options.minute = 'numeric';
+        }
+        element.removeAttribute('data-l10n-id');
+        formatter = new Intl.DateTimeFormat(navigator.languages, options);
+        element.textContent = formatter.format(this.date.shared);
```

Comment:
This is basically the same code than the future time block ahead -- I think we should factorize this to an inner function. Maybe even factorize to include the block for `dayDiff < 6` (for example using 2 functions that both call a 3rd function).

What do you think?